### ISSUE TITLE
DDP hang when using vector-quantize-pytorch within torchrun

### ISSUE THREAD

**Issue by aluminumbox** [lucidrains/vector-quantize-pytorch#105]
Hi, I've noticed that there is some distributed.all_reduce code in vector-quantize-pytorch. If I use this package in some ddp wraped code, for example, I run torchrun train.py, and train.py contains vector-quantize-pytorch code, the ddp will hang forever.

Same problem has been observed in this issue #30 

I wonder if in torchrun mode, can I just simply remove all the distributed.all_reduce and let self.kmeans_all_reduce_fn=noop, self.all_reduce_fn=noop.

Currently this is my solution and I can let the model continue training in torchrun. But I worrry that this might leads to incorrect result. Hope to know the correct way to let this package work in torchrun.

Thanks!

**Comment by aluminumbox** [comment]
> @aluminumbox hi Xiang
> 
> you can actually force ddp all reduce to be turned off by setting `sync_codebook = False`
> 
> that's interesting it hangs though, is `fairseq` being used, as in the other issue? i'm not familiar with torchrun

Thanks for the advice. I tried `sync_codebook = False` and everything is okay now. Yes when I run torch train.py that contains this package, the ddp hangs for 30mins then timeout. I am using another toolkit named `wenet`. Basically it is some code wrapped by torch.distributed and run in torchrun mode. Actually I tried to solve the problem by setting `sync_kmeans = False` earilier but it does not help. It maybe helpful to add this tip in annotation to help others who might also be stuck with torchrun.

**Comment by lucidrains** [comment]
@aluminumbox are you using gloo or nccl?

**Comment by aluminumbox** [comment]
> @aluminumbox are you using gloo or nccl?

I am using `nccl` in torchrun mutli node mode and it hangs for 30mins. 
If I use `gloo`, it gives error information.

`
terminate called after throwing an instance of 'gloo::EnforceNotMet'
  what():  [enforce fail at ../third_party/gloo/gloo/transport/tcp/pair.cc:446] op.preamble.length <= op.nbytes. 971396 vs 256
`

**Comment by lucidrains** [comment]
@aluminumbox what happens if you set `sync_codebook = True` with `nccl`?

**Comment by aluminumbox** [comment]
> @aluminumbox what happens if you set `sync_codebook = True` with `nccl`?

Sorry for the late reply. If i use `nccl`, it will hang for like 30mins than timeout, no detail error log. If I use `gloo`, it gives above error information `op.preamble.length <= op.nbytes. 971396 vs 256`

**Comment by xiaonengmiao** [comment]
I get the same issue when i train with `FSQ`. It just hangs like forever. Do we know what is the problem here?

**Comment by lucidrains** [comment]
@xiaonengmiao could you give more information on the context? is it hanging on training or eval? what version

also, try setting `quantize_dropout` to `False` and see what happens. that will help me narrow it down

**Comment by xiaonengmiao** [comment]
> @xiaonengmiao could you give more information on the context? is it hanging on training or eval? what version of vq-pytorch are you on and how you are using it
> 
> also, try setting `quantize_dropout` to `False` and see what happens. that will help me narrow it down

Thank you for your reply. When i tried to test the ddp training again on ResFSQ, this time it just works! I can not reproduce the result yesterday. Sorry for the inconvenience. I will keep testing. Right now FSQ works extremly fine. Thanks for this wonderful work!

**Comment by lucidrains** [comment]
Your welcome! Is it because you turned off the quantize dropout this time?

**Comment by lucidrains** [comment]
@xiaonengmiao anyhow do let me know if the problems comes up again and let's try a few things

FSQ is a fine quantizer, have a great time training
